File: src/unpacker.rs
use crate::{
    error::{Error, Result},
    string_finder::StringFinder,
};
use rabc::{
    abc::{
        parser::{InsIter, Instruction, Op, OpCode},
        Trait,
    },
    AbcFile, Movie,
};
use std::{collections::HashMap, io::Write};

pub struct Unpacker<'a> {
    pub movie: &'a Movie,
    pub abc: &'a AbcFile,

    pub keymap: Option<String>,
    pub methods: HashMap<u32, u8>,
    pub order: Vec<String>,
    pub binaries: HashMap<String, &'a Vec<u8>>,
}

impl<'a> Unpacker<'a> {
    pub fn new(movie: &'a Movie) -> Result<Self> {
        Ok(Unpacker {
            movie,
            abc: &movie
                .frame1()
                .ok_or(Error::missing("Movie does not have a frame1"))?
                .abcfile,
            keymap: None,
            methods: Default::default(),
            order: Default::default(),
            binaries: Default::default(),
        })
    }

    /// Unpack the movie to the given writer
    pub fn unpack(&mut self, writer: &mut dyn Write) -> Result<Option<&String>> {
        self.resolve_order()?;
        self.resolve_binaries()?;
        self.write_binaries(writer)
    }

    fn resolve_keymap(&mut self, instructions: &Vec<Instruction>) -> Result<()> {
        let mut prog = instructions.iter_prog();
        while prog.has_next() && prog.skip_until(OpCode::PushString).is_some() {
            if let Op::PushString(op) = &prog.get().op {
                self.keymap = self
                    .abc
                    .cpool
                    .strings
                    .get(op.value as usize)
                    .map(|s| s.clone());
                break;
            }
        }
        Ok(())
    }
    fn resolve_methods(&mut self) -> Result<()> {
        let keymap = self
            .keymap
            .clone()
            .ok_or(Error::missing("Keymap not found"))?;

        // Get all methods taking a ...rest argument
        // Those methods return a single character from the keymap
        for tr in &self.abc.classes[0].itraits {
            if let Trait::Method(tr) = tr {
                let method = &self.abc.methods[tr.index as usize];
                if method.need_rest() && method.max_stack == 2 {
                    let instructions = method.parse()?;
                    let mut prog = instructions.iter_prog();

                    // Get the returned character
                    if prog.skip_until(OpCode::PushByte).is_some() {
                        if let Op::PushByte(op) = &prog.get().op {
                            self.methods.insert(
                                tr.name,
                                *keymap
                                    .as_bytes()
                                    .get(op.value as usize)
                                    .ok_or(Error::IndexError())?,
                            );
                        }
                    }
                }
            }
        }
        Ok(())
    }
    pub fn resolve_order(&mut self) -> Result<()> {
        let class = self.abc.get_class(0)?;
        // Get the keymap from the cinit method
        // then resolve the methods return value
        self.resolve_keymap(&self.abc.get_method(class.cinit)?.parse()?)?;
        self.resolve_methods()?;

        let instructions = self.abc.get_method(class.iinit)?.parse()?;
        let mut prog = instructions.iter_prog();
        let mut finder = StringFinder::new(&mut prog, &self.methods);
        finder
            .prog
            .skip_until(OpCode::ConstructSuper)
            .ok_or(Error::missing("construct_super was not found"))?;

        while finder.next_string() {
            if finder.match_target("writeBytes") {
                // The next string is the binary's name
                finder.next_string();
                self.order.push(String::from_utf8(finder.build())?);
            }
        }
        Ok(())
    }
    pub fn resolve_binaries(&mut self) -> Result<()> {
        for tag in self.movie.binaries() {
            if let Some(symbol) = self.movie.symbols.get(&tag.char_id) {
                // Remove the prefix
                if let Some(name) = symbol.split_once("_").and_then(|(_, s)| Some(s)) {
                    self.binaries.insert(name.to_string(), &tag.data);
                }
            }
        }
        Ok(())
    }
    pub fn write_binaries(&self, out: &mut dyn Write) -> Result<Option<&String>> {
        for name in &self.order {
            match self.binaries.get(name) {
                Some(bin) => out.write_all(bin.as_slice())?,
                None => return Ok(Some(name)),
            }
        }
        Ok(None)
    }
}

File: lib/unpacker.cpp
#include "unpacker.hpp"
#include <cpr/cpr.h>
#include <cstring>

namespace athes::unpack {
void download(std::string url, std::vector<uint8_t>& buffer) {
    auto r = cpr::Get(
        cpr::Url { url },
        cpr::WriteCallback([&buffer](const std::string_view& data, intptr_t userdata) {
            buffer.insert(buffer.end(), data.begin(), data.end());
            return true;
        }));

    if (r.error.code != cpr::ErrorCode::OK)
        throw std::runtime_error(r.error.message);
}

Unpacker::Unpacker(std::unique_ptr<swf::StreamReader> stream)
    : buffer(), stream(std::move(stream)) {
    order    = {};
    binaries = {};
}

Unpacker::Unpacker(std::vector<uint8_t> buffer) : buffer(buffer) {
    stream   = std::make_unique<swf::StreamReader>(buffer);
    order    = {};
    binaries = {};
}

Unpacker::Unpacker(std::string url) : buffer() {
    download(url, buffer);
    stream   = std::make_unique<swf::StreamReader>(buffer);
    order    = {};
    binaries = {};
}

swf::StreamWriter Unpacker::unpack() {
    swf::StreamWriter writer;
    read_movie();
    resolve_order();
    resolve_binaries();
    write_binaries(writer);
    return writer;
}

bool Unpacker::unpack(swf::Swf& movie, std::unique_ptr<swf::StreamReader>& stream) {
    auto writer = unpack();
    if (writer.size() == 0) {
        // file was not unpacked
        // move the movie, as we didn't unpack it
        // move also the stream back, as we might still have references to the data from the movie
        movie  = std::move(this->movie);
        stream = std::move(this->stream);
        return false;
    }

    uint8_t* buffer = new uint8_t[writer.size()];
    std::memcpy(buffer, writer.get_buffer(), writer.size());
    stream = std::make_unique<swf::StreamReader>(buffer, buffer + writer.size());
    return true;
}

const size_t Unpacker::size() {
    if (!stream)
        return 0;
    return stream->size();
}

bool Unpacker::has_frame1() { return movie.abcfiles.find("frame1") != movie.abcfiles.end(); }

swf::DoABCTag* Unpacker::get_frame1() {
    if (!has_frame1())
        return nullptr;

    return movie.abcfiles.find("frame1")->second;
}

void Unpacker::read_movie() {
    if (!stream)
        throw std::runtime_error("Stream is not set.");

    movie.read(*stream);
}

void Unpacker::resolve_order() {
    if (!has_frame1())
        return;

    abc = get_frame1()->abcfile;
    Parser parser(abc->methods[abc->classes[0].cinit]);
    // Get the keymap from the cinit method
    // then resolve the methods return value
    resolve_keymap(parser.begin);
    resolve_methods();

    // Resolve the binaries order from the iinit method
    parser   = Parser(abc->methods[abc->classes[0].iinit]);
    auto ins = parser.begin;

    // Skip instructions before super()
    while (ins) {
        if (ins->opcode == OP::constructsuper)
            break;

        ins = ins->next;
    }

    std::string target = "writeBytes";
    StringFinder finder(ins);

    // Resolve binaries order
    while (finder.next_string()) {
        if (match_target(finder, target)) {
            // The next string is the binary's name
            finder.next_string();
            order.push_back(finder.build(methods));
        }
    }
}

void Unpacker::resolve_binaries() {
    if (movie.symbol_class == nullptr)
        return;

    const auto& symbols = movie.symbol_class->symbols;
    for (auto& tag : movie.binaries) {
        const auto& it = symbols.find(tag->charId);
        if (it != symbols.end()) {
            // Remove the prefix
            const auto symbol = it->second.substr(it->second.find('_') + 1);
            binaries[symbol]  = tag;
        }
    }
}

std::optional<std::string> Unpacker::write_binaries(std::ostream& file) {
    for (auto& name : order) {
        const auto& it = binaries.find(name);
        if (it == binaries.end())
            return name;

        auto data = it->second->getData();
        file.write(reinterpret_cast<const char*>(data->raw()), data->size());
    }
    return {};
}
std::optional<std::string> Unpacker::write_binaries(swf::StreamWriter& stream) {
    for (auto& name : order) {
        const auto& it = binaries.find(name);
        if (it == binaries.end())
            return name;

        stream.write(*it->second->getData());
    }
    return {};
}

bool Unpacker::match_target(StringFinder& finder, std::string& target) {
    uint32_t chr = 0;
    for (const char& c : target) {
        if (!finder.next_char(chr) || methods[chr] != c) {
            finder.skip_string();
            return false;
        }
    }
    return true;
}

void Unpacker::resolve_keymap(std::shared_ptr<Instruction> ins) {
    while (ins) {
        if (ins->opcode == OP::pushstring) {
            keymap = abc->cpool.strings[ins->args[0]];
            break;
        }
        ins = ins->next;
    }
}

void Unpacker::resolve_methods() {
    // Get all methods taking a ...rest argument
    // Those methods return a single character from the keymap
    for (auto& trait : abc->classes[0].itraits) {
        if (trait.kind == swf::abc::TraitKind::Method) {
            auto& method = abc->methods[trait.index];
            if (method.need_rest() && method.max_stack == 2) {
                Parser p(method);
                auto ins = p.begin;

                // Get the returned character
                while (ins) {
                    if (ins->opcode == OP::pushbyte) {
                        methods[trait.name] = keymap[ins->args[0]];
                        break;
                    }
                    ins = ins->next;
                }
            }
        }
    }
}
}
